### PR TITLE
Security roles

### PR DESCRIPTION
**Pull request by Olena-Pro**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 1]
```diff
@@ -31,4 +40,46 @@ public class User {
     private String lastName;
     private String shippingAddress;
     private boolean isDeleted;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]: Why did you decide to change LAZY to EAGER?

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 2]
```diff
@@ -1,29 +1,44 @@
...
-            throw new RegistrationException("User with this email: "
-                    + requestDto.email()
-                    + " already exists");
+        if (userRepository.findByEmail(requestDto.email()).isPresent()) {
+            throw new RegistrationException("Un able to complete registration");
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.password()));
+        Role role = roleRepository.findByName(Role.RoleName.USER)
+                .orElseThrow(() -> new RuntimeException("Role not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Role not found"));
```

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 3]
```diff
@@ -1,29 +1,44 @@
...
+    private final RoleRepository roleRepository;
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto)
             throws RegistrationException {
-        if (userRepository.existsByEmail(requestDto.email())) {
-            throw new RegistrationException("User with this email: "
-                    + requestDto.email()
-                    + " already exists");
+        if (userRepository.findByEmail(requestDto.email()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java]: It's better create boolean existsByEmail(String email); in UserRepo and use it here

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can't find user by email"));
```

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 5]
```diff
@@ -1,29 +1,44 @@
...
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.password()));
+        Role role = roleRepository.findByName(Role.RoleName.USER)
+                .orElseThrow(() -> new RuntimeException("Role not found"));
+        user.setRoles(Set.of(role));
         return userMapper.toDto(userRepository.save(user));
     }
+
+    @Override
+    public UserResponseDto findByEmail(String email) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java]: Why do you need this method here?

File: src/main/java/mate/academy/bookstore/controller/BookController.java [hunk 6]
```diff
@@ -39,19 +40,22 @@ public BookDto findById(@PathVariable Long id) {
...
     }
 
+    @PreAuthorize("hasAuthority('ADMIN')")
     @ResponseStatus(HttpStatus.NO_CONTENT)
     @DeleteMapping("/{id}")
     public void delete(@PathVariable Long id) {
         bookService.deleteById(id);
     }
 
+    @PreAuthorize("hasAuthority('ADMIN')")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/BookController.java:58]: Is it needed to add @PreAuthorize to other methods? Also, think about which endpoint could be called by USER and ADMIN at the same time?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/mate/academy/bookstore/model/User.java (hunk 1 above)
Comment (Olena-Pro) [reply src/main/java/mate/academy/bookstore/model/User.java]: lazy causes an error - LazyInitializationException and I don't know how to fix it

Comment (okuzan) [reply src/main/java/mate/academy/bookstore/model/User.java]: you will need to use join fetch or entity graph then

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 7]
```diff
@@ -2,4 +2,24 @@ databaseChangeLog:
...
+  - include:
+      file: /db/changelog/changes/08-set-roles-for-users.yaml
+  - include:
+      file: /db/changelog/changes/09-set-securityPassword.yaml
+
+
+
+
+
+
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: remove these lines

**Review by okuzan — CHANGES_REQUESTED** [review]
Try to fix lazyInit exception and fetch roles with something like this 
image

**Review by sarakhmen — APPROVED** [review]
LGTM!

**Pull request merged**